### ISSUE TITLE
Having trouble finding reference documentation.

### ISSUE THREAD

**Issue by lukejconnors** [reactor/reactor#638]
Hi,

I'm having trouble the old version of your reference documentation which used to be located at: https://projectreactor.io/old/reference/. I'm also having trouble finding the recent version of the reference documentation which was located at http://projectreactor.io/docs/reference/.

Have the locations of these documents been moved?

Thanks,
Luke

**Comment by simonbasle** [comment]
Yes, there was quite a lot of change, sorry for the confusion 🙇 

- the old documentation not being accessible is an overlook on our part, we'll try to fix that shortly
- the reference guide needed to be completely rewritten for 3.x, as so much has changed. This is still a work in progress though. You can see a snapshot [here](http://next.projectreactor.io/docs/core/snapshot/reference/)
- the documentation urls for 3.x have been a bit reorganized, so expect the following pattern: `/docs/{project}/{current|snapshot}/{reference|api}/` (and in the future this will allow to have urls to archives by having the version number instead of `current`/`snapshot`)

**Comment by sdwebster** [comment]
+1 this issue, but as a holdover: https://web.archive.org/web/20161024075901/http://projectreactor.io/old/reference/

**Comment by westonlast** [comment]
This is a huge indicator of the production-un-readiness of Reactor.  If you want people (like me) to take Reactor seriously, the documentation for previous versions must be officially provided, especially for major versions.  Three months later, I guess the attempt to fix this "shortly" has failed.

It was really fun to waste time figuring out that Spring Integration's Reactor support actually depends on Reactor 2: http://docs.spring.io/spring-integration/docs/4.3.9.RELEASE/reference/html/messaging-endpoints-chapter.html#_reactor_promise

There is no specification of this requirement beyond the fact that the `Promise<?>` API doesn't exist in Reactor 3.  This particular problem speaks more to the production quality of Spring Integration, inherited from optional library support.  Apparently General Availability releases don't mean much.

**Comment by smaldini** [comment]
I understand it might be an issue but the primary reason we have to delay such task is because we are primarily focused on getting a 3.1 GA that means something with Spring 5 and Reactor 3.1. Another thing we try to fix is precisely a more vertical approach that includes all spring ecosystem to anticipate such things. That's why projectreactor.io has been polished in the end, to support the first mainstream version of. Everything reactor 2 stream or promise did, reactor 3 flux or mono can in a more polished way. Reactive API is a business that is more complicated than usual given the side effects and the vocabulary of API to build, hence why even competition has to do big leaps quick (e.g. akka streams 1 -> 2 in a year, rxjava 1 -> 2 and now -> 3, ratpack, vert.x). It is not unusual in that space to find ourselves stuck in finding the right formula to address not a niche audience but a mainstream one. That's what we are doing by not only being 3 core devs on reactor-core now but also dozen of contributors from all spring projects. We do not want to reiterate the mistake we did in 1.x and 2.x but we want to make 3.x benefit from the lesson and expertise.
In the end we are also trying to migrate event bus via https://github.com/reactor/reactor-addons/compare/master...lhotari:lh-routingflux, it took time to find an adequate backpressure ready alternative/evolution to eventbus but we're going to have it too. And by then it will be due time to give a proper migration path as :
Reactor 1 is EOL.
Reactor 2 is about EOL now (3 years out)

**Comment by westonlast** [comment]
Also, class-less JAR: https://search.maven.org/#artifactdetails%7Cio.projectreactor%7Creactor%7C2.0.8.RELEASE%7CN%2FA

**Comment by smaldini** [comment]
@westonlast did you read my answer ? The fact that 3.0 is actually as far in minor/patch version as the last previous major should also speak by itself. Why not looking at http://projectreactor.io/learn to start with ? You are welcome to contribute a quick doc PR to fix that from a backported build/archive on http://github.com/reactor/projectreactor.io, or introduce an exercise of migration into our hands on / introduction https://github.com/reactor/lite-rx-api-hands-on,  or add some chapter into http://projectreactor.io/docs/core/release/reference/docs/index.html before we can actually do it ourselves.

My personal opinion if you care is that you are being offered tools for free that no one force you to use but the trade off is that we require to some extent participation from community, otherwise this would be paid products. We don't try to win the biggest share of java dev or whatever, we want to do things right especially after already acknowledging on spring.io/blog in a few tickets that Reactor 2 was rebooted in Reactor 3 from the ground as a much more focused project learning from existing branch which I spend time to try to figure out and qualifiable to be a "Spring Framework" dependency. Spring Framework in 14 years had zero fuck about deps, now it has, pretty big deal for and that's why we focus on quality and feedbacks in the new version. By comparison, Reactor 2 is mainly used in fintech and Grails.  Grails is coming with a different alternative for that now and we are working on some cool features for fintech, but not all done by reactor this time, done by a vertical of things you can pick from (reactor, spring xxx or yyy).

**Comment by westonlast** [comment]
The oft-forgotten flip-side of "free" is I get to freely criticize with no further obligation.  The situation is a mess, which you understand.  It could have been and should be handled differently.  "Free" doesn't forgo production-quality expectations nor misuse of semantic versioning. <3

**Comment by sdwebster** [comment]
@westonlast so sorry if I doomed any hope of getting those older docs back when I provided the Internet Archive link above.

Thank you @smaldini and @simonbasle so much for everything you are working on to get Reactor 3.x to a Spring-compatible version! As far as "production readiness", we're finding Reactor 2 quite serviceable in production, although it did take us some digging to learn that Grails relies on a version of Reactor not advertised on the Reactor landing page.
 
"quick doc PR to fix that from a backported build/archive" if I have time on a weekend I may look into this...I may need to ask for a couple pointers. Given how many everyday Grails users rely on Reactor for concurrency & scaling, I do think that somewhere on the home page we could also put some details about integration w/ Spring/Grails (i.e., an acknowledgement that Spring/Grails users should be looking at the Reactor 2 docs until 3.1 is accepted by Spring). I could also look into redirects for dead links...that sort of thing does go a long way towards appearing stable :)

"Grails is coming with a different alternative for that now" I'd be interested to hear more about what you mean by this Stephane...could you elaborate or link?

**Comment by smaldini** [comment]
@sdwebster I found out on https://github.com/grails/grails-core/issues/9180#issuecomment-293279486 about the grails evolution.
About what we are trying to do to port the latest feature not easily reproductible in 3.0 : 
https://github.com/reactor/reactor-addons/pull/103

Thanks for your word, I'm not saying its unfair criticism anyway but its not all doom and gloom if this is delayed for greater good and things can be found albeit with difficulties. That's what I tried to explain in substance to @westonlast.
